### PR TITLE
SMJ: fix streaming row concurrency issue for LEFT SEMI filtered join

### LINKED ISSUE #10886: SMJ producing different results than HashJoin when doing a semi join
### Describe the bug

As a part of #10728 we have added tests for semi join, and on the first merge on main they failed (they always passed). I have ignored that tests for now but that failure suggests a bug . It only happened once in several builds, and on the AMD64 architecture

```
---- fuzz_cases::join_fuzz::test_semi_join_1k_filtered stdout ----
thread 'fuzz_cases::join_fuzz::test_semi_join_1k_filtered' panicked at datafusion/core/tests/fuzz_cases/join_fuzz.rs:416:13:
assertion `left == right` failed: SortMergeJoinExec and HashJoinExec produced different row counts
  left: 100
 right: 99
stack backtrace:
   0: rust_begin_unwind
```

see https://github.com/apache/datafusion/actions/runs/9475682375/job/26107451352

### To Reproduce

https://github.com/apache/datafusion/actions/runs/9475682375/job/26107451352

### Expected behavior

Should pass

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Closes #10886 .

## Rationale for this change

There was an issue in SMJ when streamed rows handled concurrently. Race conditions for the `self.streamed_batch.join_filter_matched_idxs` lead SMJ  producing duplicates because of unexpected event ordering caused by concurrency.

## What changes are included in this PR?
Added an extra check when calculating the filter boolean mask and test methods to make fuzz test debugging easier in future

## Are these changes tested?
Yes. There is no specifc test because of unstable test nature. The issue can be reproduced on main branch with high probability by running the test 1000 times
```
#[tokio::test]
async fn test_semi_join_1k_filtered() {
   for _ in 0 .. 1000 { 
     JoinFuzzTestCase::new(
          make_staggered_batches(1000),
          make_staggered_batches(1000),
          JoinType::LeftSemi,
          Some(Box::new(col_lt_col_filter)),
      )
      .run_test(&[JoinTestType::HjSmj], false)
      .await
  }
}
```

### CODE REVIEW

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 1]
```diff
@@ -1532,17 +1532,21 @@ fn get_filtered_join_mask(
             for i in 0..streamed_indices_length {
                 // LeftSemi respects only first true values for specific streaming index,
                 // others true values for the same index must be false
-                if mask.value(i) && !seen_as_true {
+                let streamed_idx = streamed_indices.value(i);
+                if mask.value(i)
+                    && !seen_as_true
+                    && !matched_indices.contains(&streamed_idx)
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1538]: this is an actual fix!

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 2]
```diff
@@ -527,11 +550,26 @@ impl JoinFuzzTestCase {
...
+    ///         left,
+    ///         right,
+    ///         JoinType::LeftSemi,
+    ///         Some(Box::new(col_lt_col_filter)),
+    ///     )
+    ///     .run_test(&[JoinTestType::HjSmj], false)
+    ///     .await
+    /// }
+    ///
     ///     let ctx: SessionContext = SessionContext::new();
```
Comment (viirya) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:568]: The indent looks weird.

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 3]
```diff
@@ -576,6 +617,39 @@ impl JoinFuzzTestCase {
 
         println!("The data {out_name} saved as parquet into {out_path}");
     }
+
+    /// Read parquet files preserving partitions, i.e. 1 file -> 1 partition
```
Comment (viirya) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:621]: One partition is one batch here?

File: datafusion/physical-plan/src/joins/sort_merge_join.rs (hunk 1 above)
Comment (viirya) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1538]: Hmm, the fix is correct. I roughly remember I have commented on such case. `JoinType::LeftAnti` below has considered `matched_indices`. Maybe for `JoinType::LeftSem` it is missing.

**Review by viirya — APPROVED** [review]

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs (hunk 3 above)
Comment (comphead) [reply datafusion/core/tests/fuzz_cases/join_fuzz.rs:621]: correct.

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 4]
```diff
@@ -215,10 +215,6 @@ async fn test_semi_join_1k() {
     .await
 }
 
-// The test is flaky
```
Comment (alamb) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:218]: 🎉

**Pull request merged**